File: LeetCode/Easy/0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.cs
public class Solution {
    public int StrStr(string haystack, string needle) {
        if (needle.Length > haystack.Length)
        {
            return -1;
        }
        if (haystack == needle)
        {
            return 0;
        }

        var lps = BuildLps(needle);
        var n = 0;
        var i = 0;
        while (i < haystack.Length)
        {
            if (haystack[i] == needle[n])
            {
                i++;
                n++;
            }

            if (n == needle.Length)
            {
                return i - n;
                // n = lps[n - 1];
            }
            else if (i < haystack.Length && haystack[i] != needle[n])
            {
                if (n > 0)
                {
                    n = lps[n - 1];
                }
                else
                {
                    i++;
                }
            }
        }

        return -1;
    }

    private int[] BuildLps(string s)
    {
        var lps = new int[s.Length];
        var psLen = 0;
        var i = 1;

        while (i < s.Length)
        {
            if (s[i] == s[psLen])
            {
                psLen++;
                lps[i] = psLen;
                i++;
            }
            else
            {
                if (psLen > 0)
                {
                    psLen = lps[psLen - 1];
                }
                else
                {
                    lps[i] = 0;
                    i++;
                }
            }
        }

        return lps;
    }
}

File: LeetCode/Easy/0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        # using KMP algorithm, the most effective O(N + M), but the most difficult
        h_len, n_len = len(haystack), len(needle)
        def build_lps() -> List[int]:
            lps = [0] * n_len
            length = 0
            i = 1
            while i < n_len:
                if needle[i] == needle[length]:
                    length += 1
                    lps[i] = length
                    i += 1
                else:
                    if length > 0:
                        length = lps[length - 1]
                    else:
                        lps[i] = 0
                        i += 1

            return lps

        lps = build_lps()

        i = 0
        j = 0
        while i < h_len:
            if haystack[i] == needle[j]:
                i += 1
                j += 1
                if j == n_len:
                    return i - j
            else:
                if j > 0:
                    j = lps[j - 1]
                else:
                    i += 1

        return -1

        # using pointer by pointer, looks more optimal but CPython startswith is optimized
        h_len, n_len = len(haystack), len(needle)

        for i in range(h_len - n_len + 1):
            for j in range(n_len):
                if haystack[i + j] != needle[j]:
                    break
            else:
                return i
        
        return -1

        # using built-in startswith, quite nice and fast
        h_len, n_len = len(haystack), len(needle)

        for i in range(h_len - n_len + 1):
            if haystack.startswith(needle, i):
                return i
        
        return -1


        # super simple and quick basing on built-in python function, but probably it's not what we are after
        return haystack.find(needle)
        